solr/core/src/java/org/apache/solr/filestore/DistribPackageStore.java

```diff
@@ -94,8 +94,18 @@ private static Path _getRealPath(String path, Path solrHome) {
     if (!path.isEmpty() && path.charAt(0) != File.separatorChar) {
       path = File.separator + path;
     }
-    // Use concat because path might start with a slash and be incorrectly interpreted as absolute
-    return solrHome.resolve(PackageStoreAPI.PACKAGESTORE_DIRECTORY + path);
+    if (path.startsWith("\\\\")) { // Windows absolute UNC
+      throw new SolrException(BAD_REQUEST, "Illegal path " + path);
+    }
+    while (path.startsWith("/")) { // Trim all leading slashes
+      path = path.substring(1);
+    }
+    var finalPath = getPackageStoreDirPath(solrHome).resolve(path);
+    // Guard against path traversal by asserting final path is sub path of filestore
+    if (finalPath.normalize().startsWith(getPackageStoreDirPath(solrHome).normalize())) {
```

Comment:
Is normalization necessary here?

Reply:
If a path is `/foo/bar/../../baz` then the normalized is `/baz`. Probably not necessary, but I thought that if the user-input path had path traversal, then e.g. `PACKAGESTORE_PATH.resolve(../../../../foo)` would still return true for `PACKAGESTORE_PATH.startsWith()`, but if you normalize it will fail since it does not share the same root.